File: commons-core/src/main/java/xyz/wagyourtail/commons/core/reader/ParseException.java
package xyz.wagyourtail.commons.core.reader;

import org.jetbrains.annotations.NotNull;

public class ParseException extends RuntimeException implements Comparable<ParseException> {
    final int line;
    final int column;

    public ParseException(String message) {
        super(message);
        this.line = 0;
        this.column = 0;
    }

    public ParseException(String message, int line, int column) {
        super(message + getPosition(line, column));
        this.line = line;
        this.column = column;
    }

    public ParseException(String message, Throwable cause) {
        super(message, cause);
        this.line = 0;
        this.column = 0;
    }

    public ParseException(String message, int line, int column, Throwable cause) {
        super(message + getPosition(line, column), cause);
        this.line = line;
        this.column = column;
    }

    public static String getPosition(int line, int column) {
        if (column == -1) return "";
        if (line == 0) return " (at " + column + ")";
        return " (at " + line + ":" + column + ")";
    }


    @Override
    public int compareTo(@NotNull ParseException o) {
        if (line != o.line) return Integer.compare(line, o.line);
        if (column != o.column) return Integer.compare(column, o.column);
        return 0;
    }

}

File: commons-kt-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/reader/ParseException.kt
package xyz.wagyourtail.commonskt.reader

class ParseException(message: String, val line: Int = 0, val column: Int = 0, cause: Throwable? = null) : RuntimeException(message + getPosition(line, column), cause), Comparable<ParseException> {

    companion object {

        fun getPosition(line: Int, column: Int): String {
            if (column == -1) return ""
            if (line == 0) return " (at $column)"
            return " (at $line:$column)"
        }

    }

    override fun compareTo(other: ParseException): Int {
        if (line != other.line) return line.compareTo(other.line)
        if (column != other.column) return column.compareTo(other.column)
        return 0
    }

}
